### ISSUE TITLE
Compress PoV data everywhere

### ISSUE THREAD

**Issue by eskimor** [paritytech/polkadot#2574]
Right now we already compress PoV data before sending, which I believe resulted in quite significant savings of a factor of 10 or something. We could benefit from compression more, by keeping the data compressed.

Store it compressed (90% disk savings) and even build erasure encoded chunks out of the compressed data, which would improve performance for availability distribution and recovery as well, resulting in less bandwidth and again disk space savings. Only approval checkers would need to do the actual decompression, which is relatively cheap, but still.

So in summary we would save bandwidth/memory/storage everywhere and would even have less overhead with compressions/decompressions.

I might be missing something, but this sounds like an easy win in scalability.

**Comment by burdges** [comment]
As an aside, there are two parts to PoV data, the original block and the merkle proofs.  If compression compresses the merkle proofs then we should avoid the replication by optimizing how they're stored.  

This does not mean you need to treat merkle proofs specially in compressions, because compression algorithms should handle incompressible data gracefully.

**Comment by eskimor** [comment]
For what algorithm to use, Eth2 seems to go with [snappy](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/p2p-interface.md#why-are-we-using-snappy-for-compression). Which allows for examination of the decompressed size, without doing the decompression.

**Comment by bkchr** [comment]
FWIW, we already benchmarked the compression and decided on `zstd`.

**Comment by eskimor** [comment]
Considering timeouts in requests, compression would be a really good thing to have on Kusama - so not only PoVs, but also chunks and available data - would relax bandwidth requirements quite a bit. For chunks, we should do the compression before the erasure encoding obviously - and also store everything compressed in the availability store. We could make that an enum, so we can change the compression algorithm later without needing a migration.

**Comment by eskimor** [comment]
Places we want compression:

- Collations
- Availability data and by extension chunks
- PoVs
- ComittedCandidateReceipts (if large)

**Comment by bkchr** [comment]
Collations and pov are the same?

**Comment by eskimor** [comment]
Uh yeah you are right - the `Collation` struct is dead and obviously not used in the protocols.

**Comment by eskimor** [comment]
Rob already suggested, we should just make `PoV` an enum, with one entry for now:

```rust
enum PoV {
  PoVZstd(Vec<u8>)
}
```

The only place where this would need to be decompressed is at validation - as part of it. Like if decompression fails, validation fails. Sounds easy enough.

Only `ComittedCandidateReceipts` will need to be handled separately.   I like.

**Comment by eskimor** [comment]
@rphmeier works on having compressed code updates. Which makes a compressed PoV redundant in that case, so a better approach would be to have PoV payload be compressed in all cases.

**Comment by eskimor** [comment]
Instead of an enum wee can also use sp-maybe-compressed-blob here: https://github.com/paritytech/substrate/pull/8549
We pop a compress in the collation-generation subsystem, and a decompress before execution
and then remove the `CompressedPoV` type entirely.

**Comment by eskimor** [comment]
Implemented on master by @rphmeier .
